Move the version detection script from `vibe-d:tls` to `deimos/openssl`

Geod24 opened the pull request:
The OpenSSL bindings are often annoying to deal with, as different versions will expose slightly different API, which will then fail to link, and there was no good way to detect the version in use. Over the year, Vibe.d has built a script to solve this problem, which relies firstly on `pkg-config`, then on the `openssl` binary itself.

While there could probably be issues with non-conventional setups, using `pkg-config` seems like the best way forward. However, even if `vibe-d:tls` does the detection, the bindings in here still assume `1.1.0`. Moving the detection to this repository will hopefully reduce the need for duplicated work in downstream users of `openssl`.

Geod24 commented:
@CyberShadow : I suppose you will be the most affected by this, given the directory structure change, but without a dedicated source path any script would end up polluting the library.

Geod24 commented:
CC @schveiguy

schveiguy commented:
eh? I haven't had much issues with ssl. I mostly disable it and rely on the wrapper web server (apache) so far.

Geod24 commented:
@CyberShadow : Yes. You would have to:
1) Change your build system's import path to take `source` into account;
2) Call `generate_version.d` in your build system;

Point 2 will not work well on Windows, so the bindings will always assume 1.1.0. But that's no worse from what we have currently, can be overridden by the user, or fixed at the binding's level.

Geod24 commented:
@schveiguy : Just saw you in the list of "maintainers" (as in, you did a release).
The issue I'm seeing is that, using Ubuntu 22.04, OpenSSL is at v3.0.x.
Because Vibe.d depends on v1.x.x of this repo, any other library that depend on openssl must implement the same hacks as vibe-d:tls.

schveiguy commented:
If I did a release, I probably messed it up lol.

@CyberShadow is the real expert here.

CyberShadow commented:
> Call `generate_version.d` in your build system;

This isn't going to work, because I don't think `rdmd` or similar programs (`rund`, `dmd -i`) have such a facility.

What other options are there? Can we pregenerate this file for applicable versions, and/or use `version` blocks?

CyberShadow commented:
> I'm not sure what you mean ? The code depends solely on stdlib so doing `dmd -i` or `rdmd` will work.

Sorry, I meant that they don't have any hooks for running an external program before beginning the build. I understood this as that all programs which use these bindings but don't use Dub will need to add or change their build scripts to invoke this generator program, right?

> Regarding other approaches, we could have `version (OpenSSL110) { ... } else version (OpenSSL100) else import version_file;`. How does it sound ?

That sounds like the right direction, but I would go even a little further.

- In the absence of other information, assume that we are targeting the latest version of OpenSSL supported by the bindings.
- If a `version` exists, use the information from there.
- If a `version` exists which indicates that we are using a build system capable of running external commands as part of the build, then import `version_file`. I think this is a little better because all such build systems will also make it easy to specify an additional `version` to allow detecting this circumstance.

What do you think?

Geod24 commented:
> Sorry, I meant that they don't have any hooks for running an external program before beginning the build. I understood this as that all programs which use these bindings but don't use Dub will need to add or change their build scripts to invoke this generator program, right?

Correct. To be fair, if they don't have such hooks, they aren't much of a build system. But that doesn't mean we should break things for those, obviously.

> In the absence of other information, assume that we are targeting the latest version of OpenSSL supported by the bindings.

That makes sense, but at this point, it would be a breaking change. Could we do this in the next major ? Unless you plan to release a major anyway because of the files having moved ?

> If a `version` exists which specifies the OpenSSL version to target, use the information from there.

Question is, should we support every version ? I suggest to only support those listed in the "Table of content" of the [changelog](https://www.openssl.org/news/changelog.html). In practice, it means that we'll have `DeimosOpenSSL_1_1_0` meaning `1.1.0l`, and not `DeimosOpenSSL_1_1_0l`, `DeimosOpenSSL_1_1_0k`, etc...

> What do you think?

Sounds good to me.

Geod24 commented:
Updated

CyberShadow commented:
> Correct. To be fair, if they don't have such hooks, they aren't much of a build system. But that doesn't mean we should break things for those, obviously.

Right, "system" is not a good term no describe these. More like method of building sufficiently-simple D software. SSL is such an ubiquitous part of software today though that we should probably aim to keep things as simple as we can.

> That makes sense, but at this point, it would be a breaking change. Could we do this in the next major ? Unless you plan to release a major anyway because of the files having moved ?

Yeah, I think this warrants a major version bump.

> Question is, should we support every version ? I suggest to only support those listed in the "Table of content" of the [changelog](https://www.openssl.org/news/changelog.html). In practice, it means that we'll have `DeimosOpenSSL_1_1_0` meaning `1.1.0l`, and not `DeimosOpenSSL_1_1_0l`, `DeimosOpenSSL_1_1_0k`, etc...

We technically don't have the resources to actively support even one version, all support is best-effort and user-contributed. I understand that by "support" you mean "keep it in the tree", which to me raises another question - why do we aim to support multiple OpenSSL versions with one bindings version, anyway, instead of using branches? With branches we could essentially "support" any OpenSSL version indefinitely, and users can continue contributing fixes to that OpenSSL version's branch. I can see how Dub's treatment of version tags could make things a little unobvious but it doesn't seem insurmountable.

Furthermore, I'm not even sure if it makes sense to define a backwards-compatibility policy based on OpenSSL version numbers. AFAIK, OpenSSL versioning is based on compatibility with C source code, but what we really care about is compatibility with OpenSSL's ABI, which to OpenSSL developers is an implementation detail.

On a more practical note, I think we can drop code for older OpenSSL ABIs along with major version bumps of this package, and if someone would like to contribute fixes for an old ABI, we can always revive the last commit before the removal as a new branch and create tags from there.

CyberShadow commented:
> why do we aim to support multiple OpenSSL versions with one bindings version, anyway, instead of using branches?

After posting I realized that one reason in favor doing so is exactly what this pull request aims to facilitate - targeting varying OpenSSL ABI versions with one Dub configuration file. I don't think Dub has any facilities to calculate a dependency version dynamically based on an external condition, so this is as good as it gets there.

Geod24 commented:
> why do we aim to support multiple OpenSSL versions with one bindings version, anyway, instead of using branches?

I work on Ubuntu 22.04, my coworkers work on 18.04 or 20.04. The only way we can work together is to use `*` as version and not commit `dub.selections.json`. Even then, using `*` means the people on 18.04 will fetch the newer versions of the bindings.

> AFAIK, OpenSSL versioning is based on compatibility with C source code, but what we really care about is compatibility with OpenSSL's ABI, which to OpenSSL developers is an implementation detail.

Yes and no. They have taken a quite flawed approach to compatibility, but they have to maintain forward compatibility with patch versions (what they call build), otherwise shared library linking would not work. I can imagine there *might* be exceptions (e.g. if we used some private / undocumented API), but overall the public API have to be stable across "build" of the same "version".

Geod24 commented:
> We technically don't have the resources to actively support even one version

Agreed, this is more of a "best effort" basis. For reference, Vibe.d does this ATM:
https://github.com/vibe-d/vibe.d/blob/d91ee15d75a9644ad7dabe8de38db805d83af5d3/tls/vibe/stream/openssl.d#L61-L229

I simply want to encourage people to send their fixes upstream instead of hacking it downstream.

CyberShadow commented on source/deimos/openssl/opensslv.d:
```diff
@@ -1,7 +1,99 @@
+else version (DeimosOpenSSL_3_1)
+{
+    // https://www.openssl.org/news/changelog.html#openssl-31
+    // OpenSSL 3.0.3 was released 2022-05-03
+    public alias OpenSSLVersion = OpenSSLVersionTemplate!"3.0.3";
```
```suggestion
     // OpenSSL 3.1.0 was released 2022-??-??
     public alias OpenSSLVersion = OpenSSLVersionTemplate!"3.1.0";
```
?

Actually I don't see it on https://github.com/openssl/openssl/tags .

CyberShadow replied:
Yep, so why this addition?

https://github.com/D-Programming-Deimos/openssl/blob/7d8334f9d5b756d9c19a22cfbf9ee59ac4385ecf/source/deimos/openssl/opensslv.d#L54-L56

Geod24 replied:
My bad, their changelog mentions 3.1 so I originally added it, but couldn't actually find the release, so I decided to postpone it, but forgot to remove this branch.

CyberShadow commented on source/deimos/openssl/opensslv.d:
```diff
@@ -1,7 +1,99 @@
+    // It was decided in https://github.com/D-Programming-Deimos/openssl/pull/66
+    // that we should fall back to the latest supported version of the bindings,
+    // should the user provide neither explicit version nor `DeimosOpenSSLAutoDetect`
+    public alias OpenSSLVersion = OpenSSLVersionTemplate!"1.1.0h";
```
I think this should have the same version as the fallback one written by `generate_version`, so either this should be `1.1.0` or that one changed to `1.1.0h`.

CyberShadow replied:
Actually according to the comment, shouldn't this be `3.0.3`?

I think we need to establish a policy of how to determine (and when to update) the default/fallback targeted OpenSSL version.

Maybe use the OpenSSL version of `/usr/lib/libssl.so` of the latest Ubuntu LTS?

CyberShadow commented on source/deimos/openssl/opensslv.d:
```diff
@@ -28,21 +120,26 @@ import deimos.openssl._d_util;
-    OPENSSL_VERSION_PATCH   = 0,
-    OPENSSL_VERSION_BUILD   = 'h' - '`'
-}
+deprecated("Use `OpenSSLVersion.Major` instead, which has a correct value")
```
Why not make these point to the new declarations?

Geod24 replied:
I wanted to avoid misleading users. The existing names match the C names. If anyone use them, they are probably doing it wrong. I also was not envisioning a major release, which is why I wanted to avoid a change in behavior. But regardless, I think having something that matches the C macro is a bit misleading.

CyberShadow replied:
Could you please elaborate on why you think making these point to the values in the struct would be misleading? I would think that leaving them at their old values would be misleading, as they no longer represent the version of the OpenSSL that the bindings are targeting.

Geod24 replied:
If we're doing a major release, it's fine. If we were going for a minor release, it would change a constant that the user might (unwillingly / mistakenly) rely on.

CyberShadow commented on source/deimos/openssl/opensslv.d:
```diff
@@ -1,7 +1,99 @@
+    enum Text = TextVersion;
+
+    enum int Major = (Text[0] - '0');
+    static assert (Major >= 0);
```
Why e.g. `Major` and not `major`? We generally use pascal case for types in D.

Geod24 replied:
Bad habits

CyberShadow commented on source/deimos/openssl/opensslv.d:
```diff
@@ -28,21 +120,26 @@ import deimos.openssl._d_util;
 }
 
 enum OPENSSL_VERSION_NUMBER =
-    OPENSSL_MAKE_VERSION(OPENSSL_VERSION_MAJOR, OPENSSL_VERSION_MINOR, OPENSSL_VERSION_PATCH, OPENSSL_VERSION_BUILD);
+    OPENSSL_MAKE_VERSION(OpenSSLVersion.Major, OpenSSLVersion.Minor,
+                         OpenSSLVersion.Patch, OpenSSLVersion.Build);
```
It seems especially odd that this will be kept up to date, but not the individual constants. If those were to be changed to point to the struct template, then this change becomes unnecessary.

Geod24 replied:
That's a good point. I decided to update this because OpenSSL has this constant has a way to version code. With our D bindings, I don't think that this is really useful TBH.

CyberShadow replied:
I don't understand what you mean by "OpenSSL has this constant has a way to version code".

Geod24 replied:
as a way*

I meant that those were designed to work with C code and a preprocessor. Macros only make sense in a binding when they are used to alias an API. A better way to do what `OPENSSL_VERSION_NUMBER` does in D would be to implement an `opCmp` so that you could compare two versions.

But anyway, as it's the only point of contention, and I don't use those defs myself, I'll just make them alias the struct fields.

CyberShadow reviewed (commented):
Some questions:
- Is this going to interfere with caching? Such as forcing a rebuild every time?
- What happens if the version of OpenSSL is updated? If the file is generated every time, no problem, but otherwise can we make it clearer which file the user needs to delete?

> Currently, the importPath is the root directly,
> meaning that it's impossible to store D source files that
> are not directly part of the library (such as a D script).
> Moving to a source directory will allow us to add the vibe.d
> version detection script here.

I think *in theory* this should be achievable by setting `sourcePaths` (glob these files) to `deimos` and `importPaths` (tell the compiler to look here for imports) to `.`, but I don't know if Dub actually implements this distinction sufficiently meaningfully to allow this.

LGTM otherwise though haven't tried building something with this yet.

CyberShadow commented:
> Agreed, this is more of a "best effort" basis. For reference, Vibe.d does this ATM:

Right, ae has a block like that too.

Geod24 commented:
> Is this going to interfere with caching? Such as forcing a rebuild every time?
> What happens if the version of OpenSSL on the build machine is updated? If the file is generated every time, no problem, but otherwise can we make it clearer which file the user needs to delete?

The version file is generated every time, but is only written if it is different from what's on disk. So it should not interfere with build systems that take file timestamps into account, and will work if the version is updated (be it upgrade or downgrade).
It would also show up if for some reason the file ended up in CI. See https://github.com/vibe-d/vibe.d/commit/771246a03da20b30b58612d288e161f45f7186bd

Geod24 commented on dub.sdl:
```diff
@@ -19,4 +26,6 @@ configuration "unittest" {
 	targetType "executable"
 	dflags "-main"
 	excludedSourceFiles "source/deimos/openssl/applink.d"
+	preGenerateCommands `${DUB} scripts/generate_version.d` platform="posix"
```
Yes that's exactly what I was hoping for. Hoping you'll build on it once it's pushed.

Geod24 commented on scripts/generate_version.d:
```diff
@@ -0,0 +1,100 @@
+    }
+
+	string opensslVersion;
+	try {
```
Hopefully a bit more consistent now.

Geod24 commented:
Marked as resolved the thing that were addressed, and pushed an updated version.

Geod24 commented:
Updated:
- Removed explicit v3.1 branch;
- Removed deprecations on `OPENSSL_VERSION_*` and made them point to the struct fields;
- Removed `OPENSSL_VERSION_TEXT` as it includes the release date and doesn't seem generally useful;

CyberShadow reviewed (approved):
I tested building DFeed against it and it built/linked fine, just needed to update the import path symlink (`lib/deimos/openssl` from `../deimos-openssl/deimos/openssl` to `../deimos-openssl/source/deimos/openssl`).

I didn't test the Dub side but I'm guessing you have that covered.

CyberShadow commented:
Since we aim to provide compatibility with more than one OpenSSL version, I'm thinking the next tag should be just `v3.0.0` without any suffix, what do you think?

Geod24 commented:
That makes sense. And the README could probably use an update ?

CyberShadow commented:
OK, here it is:
https://github.com/D-Programming-Deimos/openssl/releases/tag/v3.0.0
